### PR TITLE
feat!: Ibis Upgrade to 5.1.0

### LINKED ISSUE #317: Support Ibis 5.0.0
Ibis 2.0 was recently released. It'd be great if we could upgrade.

Note: BigQuery backend currently blocked by ibis-bigquery package upgrade. https://github.com/ibis-project/ibis-bigquery/pull/93

### LINKED ISSUE #719: MySql - Spanner Comp-Field Validation failing for FLOAT datatype
Comp field comparison between SQL server and Big query for FLOAT datatype (column name : adjusted) failing.

Source data (SQL Server)

sr_no       adjusted                
----------- ------------------------
 2217126       7.3018260000000001
 2622064     4.8977870000000001
 3339058     43.23809
 3960098     60.973129999999998
 4585861      8.3525019999999994
 10158240     2.598392

Target data (Bigquery)

sr_no | adjusted |  
2217126 | 7.301826 |  
2622064 | 4.897787 |  
3339058 | 43.23809 |  
3960098 | 60.97313 |  
4585861 | 8.352502 |  
10158240 | 2.598392

Command used : 

data-validation validate row \
-sc MSSQL_CONN \
-tc BQ_CONN \
-tbls tempdb.dbo.dvt_prices=paras-sandbox-364713.dlp.dvt_prices \
--primary-keys sr_no \
-comp-fields symbol,adjusted \
-rr \
-rbs 5 \
--bq-result-handler paras-sandbox-364713.data_validator.results

Output : 

Screenshot 2023-02-11 at 12 33 52 AM

### LINKED ISSUE #767: validate row: Spanner TIMESTAMP Throws Exception
Test table:
```
CREATE TABLE tab_ts (
    id              INT64
,   col_datetime    TIMESTAMP
,   col_tstz        TIMESTAMP
) PRIMARY KEY (id);

INSERT INTO tab_ts
(id,col_datetime,col_tstz)
VALUES
(1,TIMESTAMP '1970-01-01 00:00:01',TIMESTAMP '1970-01-01 00:00:01-01:00');
```
Test command:
```
data-validation validate row \
-sc spanner -tc spanner \
-tbls=dvt_test.tab_ts --primary-keys=id --concat='*'
```
Exception:
```
"No matching signature for function TIMESTAMP for argument types: TIMESTAMP. Supported signatures: TIMESTAMP(STRING, [STRING]); TIMESTAMP(DATE, [STRING]) [at 73:56]\n...FORMAT_TIMESTAMP(\'%Y-%m-%d %H:%M:%S\', TIMESTAMP(`col_datetime`), \'UTC\') AS...\n"
```

When complete we need to enable the integration test for this problem. Search for "issue-767" in `tests/system/data_sources/test_spanner.py` to action this. Enabling this test is blocked by issue https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/759.

### LINKED ISSUE #814: Add Python 3.11 support
Latest version currently supported is Python 3.10

### LINKED ISSUE #821: DB2 row validation using custom query causing issue.
Running below yaml for DB2 row validation using Custom query is throwing error.

**Error** - 

AM-ERROR: Error list index out of range occured while running config file itiap_row_validation_latest.yaml. Skipping it for now.

**YAML** -

result_handler:
  project_id: rite-aid-pocs
  table_id: ra_data_validation.results
  type: BigQuery
source: DB2_CONN1
target: BQ_CONN
validations:
- calculated_fields: []
  comparison_fields:
  - cast: null
    field_alias: corporation
    source_column: itiap_corp
    target_column: corporation
  - cast: null
    field_alias: item_nbr
    source_column: itiap_itm_nbr
    target_column: item_nbr
  - cast: null
    field_alias: alternate_key_type
    source_column: itiap_alt_key_type
    target_column: alternate_key_type
  - cast: null
    field_alias: delete_ind
    source_column: itiap_dlt_ind
    target_column: delete_ind
  - cast: null
    field_alias: last_maintenance_dte
    source_column: itiap_lst_mnt_dte
    target_column: last_maintenance_dte
  - cast: null
    field_alias: last_operator_id
    source_column: itiap_lst_opr_id
    target_column: last_operator_id
  - cast: null
    field_alias: last_transaction_code
    source_column: itiap_lst_trn_cde
    target_column: last_transaction_code
  - cast: null
    field_alias: item_uom_qty
    source_column: itiap_itm_uom_qty
    target_column: item_uom_qty
  - cast: null
    field_alias: upc_type
    source_column: itiap_upc_type
    target_column: upc_type
  - cast: null
    field_alias: alternate_key_ndc
    source_column: itiap_alt_key_ndc
    target_column: alternate_key_ndc
  - cast: null
    field_alias: preferred_upc_inc
    source_column: itiap_prfd_upc_inc
    target_column: preferred_upc_inc
  - cast: null
    field_alias: preferred_upc_ind
    source_column: itiap_prfd_upc_ind
    target_column: preferred_upc_ind
  - cast: null
    field_alias: group_code_1
    source_column: itiap_grp_cde_1
    target_column: group_code_1
  - cast: null
    field_alias: group_code_2
    source_column: itiap_grp_cde_2
    target_column: group_code_2
  - cast: null
    field_alias: group_original_ind
    source_column: itiap_grp_orig_ind
    target_column: group_original_ind
  - cast: null
    field_alias: tier_id
    source_column: itiap_tier_id
    target_column: tier_id
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: alternate_key_value
    source_column: itiap_alt_key_valu
    target_column: alternate_key_value
  random_row_batch_size: null
  schema_name: null
  source_query: **SELECT itiap_corp,itiap_itm_nbr,itiap_alt_key_type,itiap_alt_key_valu,itiap_dlt_ind,itiap_lst_mnt_dte,itiap_lst_opr_id,itiap_lst_trn_cde,itiap_itm_uom_qty,itiap_upc_type,itiap_alt_key_ndc,itiap_prfd_upc_inc,itiap_prfd_upc_ind,itiap_grp_cde_1,itiap_grp_cde_2,itiap_grp_orig_ind,itiap_tier_id FROM DB2INST1.ITIAP_ITMACCS_TBL**
  table_name: null
  target_query: **SELECT corporation,item_nbr,alternate_key_type,alternate_key_value,delete_ind,last_maintenance_dte,last_operator_id,last_transaction_code,item_uom_qty,upc_type,alternate_key_ndc,preferred_upc_inc,preferred_upc_ind,group_code_1,group_code_2,group_original_ind,tier_id FROM rite-aid-pocs.ra_landing.itiap_item_access_path WHERE ra_load_dte = '2023-04-13'**
  target_schema_name: null
  target_table_name: null
  threshold: 0.0
  type: Custom-query
  use_random_rows: null

### LINKED ISSUE #832: column validation: SQL Server --sum on string column throws exception
Testing column validation with `--sum` on string columns throws an exception as below:

```
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Argument data type text is invalid for argument 1 of len function....
```

Example columns:
```
,   col_varchar_30  varchar(30)
,   col_char_2      char(2)
,   col_string      text
```

### LINKED ISSUE #833: column validation: SQL Server --sum on datetime column throws exception
Testing column validation with --sum on string columns throws an exception as below:

```
ibis.common.exceptions.OperationNotDefinedError: No translation rule for <class 'ibis.expr.operations.ExtractEpochSeconds'>
```

Example columns:
```
,   col_datetime    datetime2(3)
,   col_tstz        datetimeoffset(3)
```

### LINKED ISSUE #839: schema validation: BigQuery BIGNUMERIC
BigQuery BIGNUMERIC columns are being read by schema validation as `decimal(38, 9)`. Possibly this is because BIGNUMERIC precision and scale exceeds that supported by Ibis but I thought it should be logged anyway.

Example BigQuery column:
```
,   col_dec_38      BIGNUMERIC(38)
```

Shows in DVT schema validation as `decimal(38, 9)`.

@nehanene15 , do you think this would be resolved by the Ibis upgrade?

If we change the input type then the following integrations tests will need to be updated (search for "issue-839" in the files):

- test_hive.py
- test_oracle.py
- test_sql_server.py
- test_teradata.py

### PR DESCRIPTION
**Pull request by nehanene15**
Closes #317 (Ibis upgrade)
Closes #839 (BQ BIGNUMERIC is now expressed as decimal(76, 38))
Closes #767 (Spanner timestamp exception)
Closes #833 (MSSQL support for sum on datetime columns)
Closes #832 (MSSQL support for aggregating varchar/char columns)
Closes #719 (MSSQL/BQ column validation support for float data types)
Closes #821 (DB2 custom query index out of range bug fix)
Closes #814 (Python 3.11 support)

Breaking Change: Python 3.7 is not supported anymore. This PR also updates most of the required Python packages.

Summary: Updates Ibis to the newest 5.1.0 release. It touches a lot of files, but mainly the backend third_party connection implementations.  

Notes:
- Changed type of labels from array of structs to array of array since the Ibis upgrade interprets tuples as arrays. I.e Previously, a label would be `[(“name”, “Neha”)]` and now it will be  `[[“name”, “Neha”]]` (This doesn’t affect BQ schema or outputs to the BQ result table)
- Due to above, I had to update tests/unit/test_combiner.py to remove labels from test cases since pandas assert frame equal can’t determine equality of an array
- Snowflake support is in Alpha (has not been tested)
- The Ibis upgrade will continue to support MySQL versions < 8.0 as I overrode the new functionality that required CTEs in `third_party/ibis/ibis_mysql`

User-focused changes:
- For schema validation, previously we expressed non-nullable fields as `string[non-nullable]`. Post upgrade it will be expressed as `!string`

### CODE REVIEW

File: docs/installation.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ This tool can be natively installed on your machine or can be containerized and
 
 ## Prerequisites
 
-- Any machine with Python 3.7+ installed. Python 3.11 is not supported yet ([#814](https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/814)).
+- Any machine with Python 3.8+ installed.
```
Comment (svetakvsundhar) [inline docs/installation.md]: Could we add a note here that 3.7 is deprecated?

**Review by svetakvsundhar — CHANGES_REQUESTED** [review]
Just a minor documentation suggestion.

Also, could you grant `svetaksundhar@google.com` the `cloudbuild.builds.get` permission on `pso-kokoro-resources` so I could understand where the build is failing?

File: docs/installation.md (hunk 1 above)
Comment (nehanene15) [reply docs/installation.md]: Done!

**Comment by nehanene15** [comment]
> Just a minor documentation suggestion.
> 
> Also, could you grant `svetaksundhar@google.com` the `cloudbuild.builds.get` permission on `pso-kokoro-resources` so I could understand where the build is failing?

@svetakvsundhar You should have access now!

**Comment by svetakvsundhar** [comment]
thanks @nehanene15 ! It seems like something wrong with the integration test, but I can't see the log specifically. Does it pass locally?

**Comment by nehanene15** [comment]
> thanks @nehanene15 ! It seems like something wrong with the integration test, but I can't see the log specifically. Does it pass locally?

@svetakvsundhar It's all passing now - there were some syntax fixes specific to Python 3.8 and the package upgrades

**Review by svetakvsundhar — APPROVED** [review]
Thanks Neha! Looks great!!

File: data_validation/clients.py [hunk 2]
```diff
@@ -59,50 +45,45 @@ def get_client_call(*args, **kwargs):
...
     )
-    ibis_client = ibis_bigquery.connect(
-        project_id, dataset_id=dataset_id, credentials=credentials
+
+    ibis_client = ibis.bigquery.connect(
+        project_id=project_id, dataset_id=dataset_id, credentials=credentials
     )
 
-    # Override the BigQuery client object to ensure the correct user agent is
+    # # Override the BigQuery client object to ensure the correct user agent is
```
Comment (nj1973) [inline data_validation/clients.py]: Is this extra `#` a mistake?

File: data_validation/clients.py [hunk 3]
```diff
@@ -143,15 +124,15 @@ def get_ibis_table(client, schema_name, table_name, database_name=None):
...
     database_name (str): Database name (generally default is used)
     """
-    if type(client) in [
-        OracleClient,
-        PostgreSQLClient,
-        DB2Client,
-        MSSQLClient,
-        RedShiftClient,
+    if client.name in [
+        "oracle",
```
Comment (nj1973) [inline data_validation/clients.py:128]: Should these literals be defined as constants seeing as they are used in multiple places?

File: docs/connections.md [hunk 4]
```diff
@@ -189,6 +189,10 @@ data-validation connections add
 ```
 
 ## MySQL
+DVT only supports MySQL version 8.0+. 
+
+TODO: link latest version before Ibis upgrade
```
Comment (nj1973) [inline docs/connections.md]: Is this TODO still a todo or should it have been actioned by this point?

File: third_party/ibis/ibis_cloud_spanner/datatypes.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2021 Google Inc.
+# Copyright 2023 Google Inc.
```
Comment (nj1973) [inline third_party/ibis/ibis_cloud_spanner/datatypes.py]: I seem to recall we are not to changes dates in copyright headers?

File: third_party/ibis/ibis_mssql/datatypes.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+def _datetimeoffset(_, sa_type, nullable=True):
+    return dt.Timestamp(timezone="UTC", nullable=nullable)
+
+
+@dt.dtype.register(MSDialect, mssql.DATETIME2)
+def _datetime2(_, sa_type, nullable=True):
+    return dt.Timestamp(nullable=nullable)
+
+
+# Needs to be VARCHAR insteaad of NVARCHAR for Hash function
```
Comment (nj1973) [inline third_party/ibis/ibis_mssql/datatypes.py]: Typo: insteaad

File: third_party/ibis/ibis_oracle/datatypes.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+from sqlalchemy.dialects import oracle
+from sqlalchemy.dialects.oracle.cx_oracle import OracleDialect_cx_oracle
+
+import ibis.expr.datatypes as dt
+from ibis.backends.base.sql.alchemy.datatypes import ibis_type_to_sqla
+import cx_Oracle
+
+# Update to avoid cast to CLOB/Text
+ibis_type_to_sqla[dt.String] = sa.sql.sqltypes.String(length=4000)
+ibis_type_to_sqla[dt.Float32] = sat.Float(precision=23).with_variant(oracle.FLOAT(), 'oracle')
```
Comment (nj1973) [inline third_party/ibis/ibis_oracle/datatypes.py]: I'd be interested to know the purpose of the `with_variant()` call, did this code come from the Ibis project or is it our own?

File: third_party/ibis/ibis_oracle/datatypes.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    typename = col[1]
+    typ = _type_mapping.get(typename)
+    if typ is None:
+        raise NotImplementedError(
+            f"Oracle type {typename} is not supported"
+        )
+    
+    if typename == cx_Oracle.DB_TYPE_NUMBER:
+        if col[4] == 0 and col[5] == -127:
+            # This will occur if type is NUMBER with no precision/scale
```
Comment (nj1973) [inline third_party/ibis/ibis_oracle/datatypes.py:56]: Oracle NUMBER without precision or scale is not equivalent to NUMBER(38,0). Did this code come through from the Ibis project?

File: data_validation/clients.py (hunk 3 above)
Comment (nj1973) [reply data_validation/clients.py:128]: Or did we leave them as string literals because are defined in Ibis itself and therefore not really under our control?

**Review by nj1973 — CHANGES_REQUESTED** [review]
All minor comments.

File: third_party/ibis/ibis_addon/operations.py [hunk 9]
```diff
@@ -157,201 +125,222 @@ def strftime_bigquery(translator, expr):
...
     sa_arg = t.translate(arg)
-    sa_type = t.get_sqla_type(typ)
-
-    # Specialize going from an integer type to a timestamp
-    if isinstance(arg.type(), dt.Integer) and isinstance(sa_type, sa.DateTime):
-        return sa.func.timezone('UTC', sa.func.to_timestamp(sa_arg))
 
     # Specialize going from numeric(p,s>0) to string
-    if isinstance(arg.type(), dt.Decimal) and arg.type().scale > 0 and typ.equals(dt.string):
+    if arg_dtype.is_decimal() and arg_dtype.scale > 0 and typ.equals(dt.string):
```
Comment (nj1973) [inline third_party/ibis/ibis_addon/operations.py]: I created a PostgreSQL `decimal` column with no precision or scale and get an error from this line:

```
  File "/usr/local/google/home/neiljohnson/github/professional-services-data-validator/third_party/ibis/ibis_addon/operations.py", line 260, in sa_cast_postgres
    if arg_dtype.is_decimal() and arg_dtype.scale > 0 and typ.equals(dt.string):
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

On the `develop` branch I don't get the exception. But it seems to be unrelated to this line, on develop a scale of `None` becomes `0`. That isn't happening on the feature branch.

File: data_validation/clients.py (hunk 2 above)
Comment (nehanene15) [reply data_validation/clients.py]: yup - good catch!

File: data_validation/clients.py (hunk 3 above)
Comment (nehanene15) [reply data_validation/clients.py:128]: Yeah they're defined in Ibis itself - the new upgrade uses a Backend class instead of a Client class, and each Backend has a name property indicating the DB type.

File: docs/connections.md (hunk 4 above)
Comment (nehanene15) [reply docs/connections.md]: Yeah, will delete this since we will support MySQL < 8.0

File: third_party/ibis/ibis_cloud_spanner/datatypes.py (hunk 5 above)
Comment (nehanene15) [reply third_party/ibis/ibis_cloud_spanner/datatypes.py]: Oh yeah - I'll revert this. I only meant to add 2023 headers to new files

File: third_party/ibis/ibis_oracle/datatypes.py (hunk 7 above)
Comment (nehanene15) [reply third_party/ibis/ibis_oracle/datatypes.py]: The upgrade to SQLalchemy 2.0 required this call to prevent the following exception: `sqlalchemy.exc.ArgumentError: Oracle FLOAT types use 'binary precision', which does not convert cleanly from decimal 'precision'.`

[Here](https://docs.sqlalchemy.org/en/20/dialects/oracle.html#sqlalchemy.dialects.oracle.BINARY_DOUBLE.params.precision) is the alchemy docs link as well.

File: third_party/ibis/ibis_oracle/datatypes.py (hunk 8 above)
Comment (nehanene15) [reply third_party/ibis/ibis_oracle/datatypes.py:56]: I thought the default would be NUMBER(38,0) based on this: https://stackoverflow.com/questions/593197/what-is-the-default-precision-and-scale-for-a-number-in-oracle

I used 38 as the precision to account for the maximum given values stored in the column. Open to change this though - what would the default precision/scale be?

File: third_party/ibis/ibis_addon/operations.py (hunk 9 above)
Comment (nehanene15) [reply third_party/ibis/ibis_addon/operations.py]: Looks like the upgrade added the default scale of 'None' which I'll have to handle: https://github.com/ibis-project/ibis/blob/b9977fbe3ed30033fc2500abe061eda8d7e26eed/ibis/expr/datatypes/core.py#L669 

Previously, the scale had to be an int so we didn't get an exception

File: third_party/ibis/ibis_oracle/datatypes.py (hunk 8 above)
Comment (nj1973) [reply third_party/ibis/ibis_oracle/datatypes.py:56]: If precision is specified, e.g. NUMBER(12) then scale defaults to 0 in the data dictionary. We shouldn't need to worry about these. If precision and scale are excluded then there is no default, Oracle does not enforce a fixed decimal place and any value can be stored. I believe PostgreSQL works in the same way.

I've tested a NUMBER(*) column on the `develop` branch and the same happens there. Therefore I think this is a problem for another day. I'll mark this conversation as resolved.

File: third_party/ibis/ibis_oracle/datatypes.py (hunk 7 above)
Comment (nj1973) [reply third_party/ibis/ibis_oracle/datatypes.py]: Oracle has it's own dedicated float-32 (BINARY_FLOAT) and float-64 (BINARY_DOUBLE) data types so I think they would be more appropriate to use. But I don't fully understand what `ibis_type_to_sqla` is used for so I'm happy for this conversation to be resolved seeing as integration tests are passing.

File: third_party/ibis/ibis_addon/operations.py (hunk 9 above)
Comment (nj1973) [reply third_party/ibis/ibis_addon/operations.py]: Now I've slept on it I think `None` is correct. This line of code (which I added as it happens) is probably not defensive enough. Perhaps it should be `(arg_dtype.scale or 0) > 0`?

Comment (nj1973) [reply third_party/ibis/ibis_addon/operations.py]: I've tested Oracle `NUMBER` vs PostgreSQL `decimal`, both without precision or scale.

On `develop` we get:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 2cae2db8-8dd8-4628-8b40-2db469a9ae1c │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal(38, 0)     │ decimal(38, 0)     │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```

On this feature branch we get:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ 80f8ecbc-b090-488e-a089-733766876bcf │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal(38, 0)     │ decimal            │ fail                │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```

In a purists world I think `decimal` with no precision or scale is the correct outcome and therefore the correct action would be to remove the defaults from Oracle. But if this branch should reflect what we had before then perhaps PostgreSQL should have the same (incorrect) defaults as it had on `develop`.

Comment (nehanene15) [reply third_party/ibis/ibis_addon/operations.py]: I went ahead and removed the defaults from Oracle. I also added the extra condition in this line to `if arg_dtype.scale and arg_dtype.scale > 0` to account for scale being None.

Comment (nj1973) [reply third_party/ibis/ibis_addon/operations.py]: I've retested and that works as expected:

```
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
│ a845a755-eb50-4ad9-ac63-d582a5dc527f │ Schema            │ Schema            │ []       │ dvt_test.dvt_ora2pg_types │ dvt_test.dvt_ora2pg_types │ col_num              │ col_num              │ decimal            │ decimal            │ success             │
├──────────────────────────────────────┼───────────────────┼───────────────────┼──────────┼───────────────────────────┼───────────────────────────┼──────────────────────┼──────────────────────┼────────────────────┼────────────────────┼─────────────────────┤
```

**Review by nj1973 — COMMENTED** [review]
Reviewed code and did custom testing of Oracle to PostgreSQL: LGTM

**Review by nj1973 — APPROVED** [review]
LGTM

File: third_party/ibis/ibis_cloud_spanner/__init__.py [hunk 10]
```diff
@@ -1,2 +1,220 @@
...
+
+    def dataset(self, database):
+        spanner_client = spanner.Client()
+        instance = spanner_client.instance(self.data_instance)
+        database = instance.database(database)
+
+    def exists_database(self, name):
+        return self.instance.database(name).exists()
+
+    def create_table():
```
Comment (renzokuken) [inline third_party/ibis/ibis_cloud_spanner/__init__.py:197]: are these stubbed functions required?

File: third_party/ibis/ibis_postgres/client.py [hunk 11]
```diff
@@ -12,98 +12,37 @@
...
-
-    brackets = spaceless(LBRACKET).then(spaceless(RBRACKET))
-
-    pg_array = parsy.seq(decimal, brackets.at_least(1).map(len)).combine(
-        lambda value_type, n: toolz.nth(n, toolz.iterate(dt.Array, value_type))
-    )
-
-    ty = pg_array | decimal
-    return ty.parse(text)
+def _metadata(self, query: str) -> sch.Schema:
```
Comment (renzokuken) [inline third_party/ibis/ibis_postgres/client.py:24]: What's the purpose of this _metadata method? What is it extending from the base client?

**Review by renzokuken — APPROVED** [review]
LGTM - I have a couple of questions but overall this looks like it will simplify things in the long run.

File: third_party/ibis/ibis_oracle/__init__.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+    def _metadata(self, query)  -> Iterable[Tuple[str, dt.DataType]]:
+        if (
+            re.search(r"^\s*SELECT\s", query, flags=re.MULTILINE)
+            is not None
+        ):
+            query = f"({query})"
+
+        with self.begin() as con:
+            result = con.exec_driver_sql(f"SELECT * FROM {query} t0 WHERE ROWNUM <= 1")
```
Comment (nj1973) [inline third_party/ibis/ibis_oracle/__init__.py:76]: I think we are missing parentheses around the sub-query here. When testing custom-query I get:

```
sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-00903: invalid table name
[SQL: SELECT * FROM select * from dvt_test.dvt_ora2pg_types t0 WHERE ROWNUM <= 1]
```

**Review by nj1973 — CHANGES_REQUESTED** [review]

File: third_party/ibis/ibis_cloud_spanner/__init__.py (hunk 10 above)
Comment (nehanene15) [reply third_party/ibis/ibis_cloud_spanner/__init__.py:197]: Yeah - the Ibis Backend function requires those methods to be present otherwise it'll throw an Exception

File: third_party/ibis/ibis_postgres/client.py (hunk 11 above)
Comment (nehanene15) [reply third_party/ibis/ibis_postgres/client.py:24]: The metadata method is mainly used for custom query validations for our purpose. It takes the custom query and returns the schema associated with the results of the query.

File: third_party/ibis/ibis_oracle/__init__.py (hunk 12 above)
Comment (nehanene15) [reply third_party/ibis/ibis_oracle/__init__.py:76]: Yeah - I think this is a result of me mistakenly deleting the re.IgnoreCase flag on [this line](https://github.com/GoogleCloudPlatform/professional-services-data-validator/pull/894/files/0ff3aaa42e769cb0a05e270cea3a85cae00dad90#diff-1b5e7ce512e4c118ff2bc62fff9cc2940a08c90e9a4affee9cc1f7c116449bfcR69) which would add the parentheses around the sub-query. Fixing this

**Review by nj1973 — APPROVED** [review]
LGTM, awesome job.

**Comment by piyushsarraf** [comment]
Hi
I was testing the ibis-upgrade branch and saw some errors wrt to BQ and RedShift. 
I was doing a BQ - BQ row validation and got a error of 2 packages found for backend 'bigquery'
image
And for RedShift I was doing a row validation and the error seems to be with version of SQL Alchemy.
image

**Comment by nj1973** [comment]
@piyushsarraf , I also saw the "2 packages found for backend 'bigquery'" message. It was because my virtual environment was not clean. I fixed it by removing the venv, creating it fresh and then `pip` installing DVT.

**Comment by nehanene15** [comment]
@piyushsarraf 
As of now the only blocker to merge the PR is the Redshift unrecognized configuration parameter. It could be because I upgraded the `psycopg2-binary` package from 2.9.3 to 2.9.6, or due to SQLAlchemy conflicts similar to this - https://stackoverflow.com/questions/76532906/unrecognized-configuration-parameter-standard-conforming-strings-while-queryin

**Comment by Raniksingh** [comment]
@nehanene15 I just verified the below version which works for Redshift. 
psycopg2-binary ==2.9.6 and  SQLAlchemy ==1.4.39

**Comment by sundar-mudupalli-work** [comment]
Hey Neha,
I was excited that Teradata query worked with DVT, and looked into getting generate-partitions to work on Teradata. Good news - I have it working and tested on oracle, sqlserver, bigquery, mysql, postgres, hive and teradata. Test functions for these have been written as well. Not so Good news - All this work is on the ibis-upgrade branch - it is all in my local repository. It is only a few files (4-5), so I can hack it into any branch. Or I need someone to tell me how to merge it into the ibis-upgrade branch without messing other things. Now realize that I should have created another branch off ibis-upgrade.

Thanks.

Sundar Mudupalli
